src/ICeWall/src/ICeWall.tcl

```diff
@@ -3561,10 +3713,914 @@ namespace eval ICeWall {
+      # debug "end: physical only padcell"
+      return $padcell
+    }
+
+    if {[set signal [$block findNet $signal_name]] == "NULL"} {
+      set alt_signal_name 0
+      if {[dict exists $padcell pad_pin_name]} {
+        set alt_signal_name 1
+        set try_signal [format [dict get $padcell pad_pin_name] $signal_name]
+        set signal [$block findNet $try_signal]
+      } elseif {[dict exists $footprint pad_pin_name]} {
+        set alt_signal_name 1
+        set try_signal [format [dict get $footprint pad_pin_name] $signal_name]
+        set signal [$block findNet $try_signal]
+      }
+
+      if {$signal == "NULL"} {
+        if {[$block findInst $signal_name] == "NULL"} {
+          if {$alt_signal_name == 1} {
+            utl::error PAD 92 "No signal $signal_name or $try_signal defined for padcell [dict get $padcell name]"
+          } else {
+            utl::error PAD 91 "No signal $signal_name defined for padcell [dict get $padcell name]"
+          }
+        } else {
+          dict set padcell use_signal_name $signal_name
+          # debug "end: inst name as signal name padcell"
+          return $padcell
+        }
+      } else {
+        dict set padcell use_signal_name $try_signal
+        # debug "end: alt signal name padcell"
+        return $padcell
+      }
+    }
+
+    utl::error PAD 93 "Signal \"$signal_name\" not found in design "
+  }
+
+  proc check_edge_name {edge_name} {
+    set edge_names {
+      top {top TOP Top T t}
+      bottom {bottom BOTTOM Bottom B b}
+      right {right RIGHT Right R r}
+      left {left LEFT Left L l}
+    }
+
+    dict for {edge options} $edge_names {
+      if {[lsearch $options $edge_name] > -1} {
+        return $edge
+      }
+    }
+    utl::error PAD 94 "Value for -edge_name ($edge_name) not permitted, choose one of bottom, right, top or left"
+  }
+
+  proc check_cell_type {type} {
+    variable library
+
+    if {[dict exists $library types]} {
+      if {[dict exists $library types $type]} {
+        return $type
+      } else {
+        utl::error PAD 95 "Value for -type ($type) does not match any library types ([dict keys [dict get $library types]])"
+      }
+    } else {
+      utl::error PAD 98 "No library types defined"
+    }
+  }
+
+  proc check_xy {xy} {
+    # debug $xy
+    if {[llength $xy] != 4} {
+      error "expecting a 4 element list in the form \"x number y number\""
+    }
+    if {![dict exists $xy x]} {
+      error "no value specified for x"
+    }
+    if {![dict exists $xy y]} {
+      error "no value specified for y"
+    }
+    if {![is_number [dict get $xy x]]} {
+      error "x co-ordinate ([dict get $xy x]), not recognised as a nummber"
+    }
+    if {![is_number [dict get $xy y]]} {
+      error "y co-ordinate ([dict get $xy y]), not recognised as a nummber"
+    }
+  }
+
+  proc check_rowcol {rowcol} {
+    if {[llength $rowcol] != 4} {
+      error "expecting a 4 element list in the form \"row <integer> col <integer>\""
```

Comment:
should these be utl::error? or was this intentional? (same for all of check_xy and remaining checks in check_rowcol)

Reply:
This was intentional - it started out with the check_xy function - I wanted to have a singe function that checked the syntax of the xy setting, but I also wanted the error message that is fed back to the user to be specify about whether it was for center or origin. All calls of check_xy are wraped in a catch, so that if the function call to check_xy throws an error, then it is caught in the code that knows more about the context of the call to check_xy and can give a better error message.

I took the same approach with check_rowcol, but looking back at it now, it seems like it was a bit of overkill, as the code wound up with just a single call to check_rowcol.